### PR TITLE
Modernize plasma store (C to C++ changes).

### PR DESCRIPTION
**Pull request by pcmoritz**
Changes include:

- The plasma store is now a class
- A type safe version of the event loop
- Less context being passed around due to C++ closures

### CODE REVIEW

File: src/common/io.cc [hunk 1]
```diff
@@ -384,6 +384,36 @@ int64_t read_buffer(int fd, int64_t *type, UT_array *buffer) {
   return 0;
 }
 
+int64_t read_vector(int fd, int64_t *type, std::vector<uint8_t> &buffer) {
+  int64_t version;
+  int closed = read_bytes(fd, (uint8_t *) &version, sizeof(version));
```
Comment (robertnishihara) [inline src/common/io.cc]: We probably shouldn't be using C-style casts

File: src/plasma/plasma_manager.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #define PLASMA_MANAGER_H
 
 #include <poll.h>
+#include "uthash.h"
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.h]: Do we need this?

File: src/plasma/plasma_store.h [hunk 3]
```diff
@@ -1,103 +1,167 @@
...
+   *
+   * @return Void.
+   */
+  void subscribe_to_updates(int client_fd);
+
+  /**
+   * Connect a new client to the PlasmaStore.
+   *
+   * @param listener_sock The socket that is listening to incoming connections.
+   *
```
Comment (robertnishihara) [inline src/plasma/plasma_store.h]: remove this line

File: src/plasma/plasma_store.h [hunk 4]
```diff
@@ -1,103 +1,167 @@
...
+   * @param client_fd The file descriptor of the client making this request.
+   * @param Void.
+   */
+  void release_object(ObjectID object_id, int client_fd);
+
+  /**
+   * Subscribe a file descriptor to updates about new sealed objects.
+   *
+   * @param client_fd The file descriptor of the client making this request.
+   *
```
Comment (robertnishihara) [inline src/plasma/plasma_store.h]: remove this line

File: src/plasma/plasma_store.h [hunk 5]
```diff
@@ -1,103 +1,167 @@
...
+   * @return Void.
+   */
+  void connect_client(int listener_sock);
+
+  /**
+   * Disconnect a client from the PlasmaStore.
+   *
+   * @param client_fd The file descriptor of the client that is disconnected.
+   *
```
Comment (robertnishihara) [inline src/plasma/plasma_store.h]: remove this line

File: src/plasma/plasma_store.h [hunk 6]
```diff
@@ -1,103 +1,167 @@
...
+                    int64_t metadata_size,
+                    int client_fd,
+                    PlasmaObject *result);
+
+  /**
+   * Delete objects that have been created in the hash table. This should only
+   * be called on objects that are returned by the eviction policy to evict.
+   *
+   * @param object_ids Object IDs of the objects to be deleted.
+   *
```
Comment (robertnishihara) [inline src/plasma/plasma_store.h]: remove this line

File: src/plasma/eviction_policy.cc [hunk 7]
```diff
@@ -19,8 +19,8 @@ class LRUCache {
   void add(const ObjectID &key, int64_t size) {
     auto it = item_map_.find(key);
     CHECK(it == item_map_.end());
-    item_list_.push_front(std::make_pair(key, size));
-    item_map_.insert(std::make_pair(key, item_list_.begin()));
+    item_list_.emplace_front(key, size);
+    item_map_.emplace(key, item_list_.begin());
```
Comment (robertnishihara) [inline src/plasma/eviction_policy.cc]: looks like this should work since we're using lists, but maybe add a comment that this relies on the fact that we're using lists so the iterators are not invalidated

File: src/plasma/eviction_policy.cc [hunk 8]
```diff
@@ -43,114 +43,71 @@ class LRUCache {
...
-  LRUCache cache;
-};
-
-EvictionState *EvictionState_init() {
-  EvictionState *s = new EvictionState();
-  s->memory_used = 0;
-  return s;
-}
+EvictionPolicy::EvictionPolicy(PlasmaStoreInfo *store_info)
+    : memory_used_(0), store_info_(store_info), cache_(new LRUCache()) {}
```
Comment (robertnishihara) [inline src/plasma/eviction_policy.cc]: maybe make `cache_` a non-pointer? so we can avoid calling new/delete

File: src/plasma/plasma_store.cc [hunk 9]
```diff
@@ -48,27 +44,11 @@ void dlfree(void *);
 size_t dlmalloc_set_footprint_limit(size_t bytes);
 }
 
-/** Contains all information that is associated with a Plasma store client. */
-struct Client {
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Why remove the client struct? Don't we want to keep around the struct (even if it's only wrapping a file descriptor at the moment)?

**Review by robertnishihara — COMMENTED** [review]
Still in the middle of reviewing

File: src/plasma/plasma_store.cc [hunk 10]
```diff
@@ -340,33 +239,36 @@ void return_from_get(PlasmaStoreState *store_state, GetRequest *get_req) {
    * tables if it is present there. It should only be present there if the get
    * request timed out. */
   for (ObjectID &object_id : get_req->object_ids) {
-    remove_get_request_for_object(store_state, object_id, get_req);
+    auto &get_requests = object_get_requests_[object_id];
+    get_requests.erase(
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: This line with the `std::remove` inside of the `erase` is hard to parse. Can you add a comment explaining what is happening here? Is there a way to make it clearer? Note that it used to be a whole separate function.

**Comment by robertnishihara** [comment]
This makes a bunch of progress on #406.

File: src/plasma/plasma_manager.h (hunk 2 above)
Comment (pcmoritz) [inline src/plasma/plasma_manager.h]: Yeah, the plasma manager is still using uthash and now that plasma.h is not including it any more, we need to include it here.

File: src/common/io.cc (hunk 1 above)
Comment (pcmoritz) [inline src/common/io.cc]: I fixed it in the next PR, which modernizes some of the IO routines that the plasma store uses.

File: src/plasma/plasma_store.h (hunk 6 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.h]: done

File: src/plasma/plasma_store.h (hunk 4 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.h]: done

File: src/plasma/plasma_store.h (hunk 3 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.h]: done

File: src/plasma/plasma_store.h (hunk 5 above)
Comment (pcmoritz) [inline src/plasma/plasma_store.h]: done

File: src/plasma/eviction_policy.cc (hunk 7 above)
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/eviction_policy.cc (hunk 8 above)
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/plasma_events.h [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+  EventLoop();
+
+  /**
+   * Add a new file event handler to the event loop.
+   *
+   * @param fd The file descriptor we are listening to.
+   * @param events The flags for events we are listening to (read or write).
+   * @param callback The callback that will be called when the event happens.
+   *
```
Comment (pcmoritz) [inline src/plasma/plasma_events.h]: done

File: src/plasma/plasma_events.h [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+   * @return Returns true if the event handler was added successfully.
+   */
+  bool add_file_event(int fd, int events, FileCallback callback);
+
+  /**
+   * Remove a file event handler from the event loop.
+   *
+   * @param fd The file descriptor of the event handler.
+   *
```
Comment (pcmoritz) [inline src/plasma/plasma_events.h]: done

File: src/plasma/plasma_events.h [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+   * @return Void.
+   */
+  void remove_file_event(int fd);
+
+  /** Register a handler that will be called after a time slice of
+   *  "timeout" milliseconds.
+   *
+   *  @param timeout The timeout in milliseconds.
+   *  @param callback The callback for the timeout.
+   *
```
Comment (pcmoritz) [inline src/plasma/plasma_events.h]: done

File: src/plasma/plasma_events.h [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+                                  int fd,
+                                  void *context,
+                                  int events);
+
+  static int timer_event_callback(aeEventLoop *loop,
+                                  long long timer_id,
+                                  void *context);
+
+  aeEventLoop *loop_;
+  std::unordered_map<int, FileCallbackData *> file_callbacks_;
```
Comment (pcmoritz) [inline src/plasma/plasma_events.h]: done

File: src/plasma/eviction_policy.cc [hunk 15]
```diff
@@ -43,114 +43,71 @@ class LRUCache {
...
 }
 
-void EvictionState_object_created(EvictionState *eviction_state,
-                                  PlasmaStoreInfo *plasma_store_info,
-                                  ObjectID object_id) {
-  ObjectTableEntry *entry = plasma_store_info->objects[object_id];
-  eviction_state->cache.add(object_id,
-                            entry->info.data_size + entry->info.metadata_size);
+void EvictionPolicy::object_created(ObjectID object_id) {
+  auto entry = store_info_->objects[object_id].get();
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: see this: http://stackoverflow.com/questions/17536731/shared-ptr-is-to-weak-ptr-as-unique-ptr-is-to-what

In practice people seem to use plain pointers as weak pointers for std::unique_ptr, so that's what we are doing here.

File: src/plasma/plasma_store.cc [hunk 16]
```diff
@@ -377,186 +279,161 @@ void update_object_get_requests(PlasmaStoreState *store_state,
...
-      delete_object(plasma_state, objects_to_evict[i]);
-    }
-    /* Free the array of objects to evict. This array was originally allocated
-     * by the eviction policy. */
-    free(objects_to_evict);
-  }
+void PlasmaStore::connect_client(int listener_sock) {
+  int new_socket = accept_client(listener_sock);
+  /* Add a callback to handle events on this socket. */
+  loop_->add_file_event(new_socket, kEventLoopRead,
```
Comment (pcmoritz) [inline src/plasma/plasma_store.cc]: done

File: src/plasma/plasma_store.cc [hunk 17]
```diff
@@ -580,11 +457,13 @@ void send_notifications(event_loop *loop,
        * there is room in the socket's send buffer. Callbacks can be added
        * more than once here and will be overwritten. The callback is removed
        * at the end of the method. */
-      event_loop_add_file(plasma_state->loop, client_sock, EVENT_LOOP_WRITE,
-                          send_notifications, plasma_state);
+      loop_->add_file_event(client_fd, kEventLoopWrite,
```
Comment (pcmoritz) [inline src/plasma/plasma_store.cc]: done

File: src/plasma/plasma_store.cc [hunk 18]
```diff
@@ -776,14 +622,15 @@ void start_server(char *socket_name, int64_t system_memory) {
...
-  PlasmaStoreState *state = new PlasmaStoreState(loop, system_memory);
+  EventLoop loop;
+  PlasmaStore store(&loop, system_memory);
   int socket = bind_ipc_sock(socket_name, true);
   CHECK(socket >= 0);
-  event_loop_add_file(loop, socket, EVENT_LOOP_READ, new_client_connection,
-                      state);
-  g_state = state;
-  event_loop_run(loop);
+  loop.add_file_event(socket, kEventLoopRead,
```
Comment (pcmoritz) [inline src/plasma/plasma_store.cc]: done

File: src/plasma/plasma_store.cc [hunk 19]
```diff
@@ -340,33 +239,37 @@ void return_from_get(PlasmaStoreState *store_state, GetRequest *get_req) {
    * tables if it is present there. It should only be present there if the get
    * request timed out. */
   for (ObjectID &object_id : get_req->object_ids) {
-    remove_get_request_for_object(store_state, object_id, get_req);
+    auto &get_requests = object_get_requests_[object_id];
+    /* Erase get_req from vector, see http://stackoverflow.com/a/16013546 */
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: I see what you're doing here now. This idiom is used to remove *all* instances of `get_req` from the `get_requests` vector. However, since `get_req` can only occur once in the vector (we were relying on this fact before), this could also be done with 

```
get_requests.erase(std::find(get_requests.begin(), get_requests.end(), get_req))
```

which is probably a bit cleaner.

File: src/plasma/plasma_events.cc [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+  FileCallback *callback = reinterpret_cast<FileCallback *>(context);
+  (*callback)(events);
+}
+
+int EventLoop::timer_event_callback(aeEventLoop *loop,
+                                    long long timer_id,
+                                    void *context) {
+  TimerCallback *callback = reinterpret_cast<TimerCallback *>(context);
+  (*callback)(timer_id);
+}
```
Comment (robertnishihara) [inline src/plasma/plasma_events.cc]: This needs to return an int, right?

File: src/plasma/eviction_policy.cc [hunk 21]
```diff
@@ -1,156 +1,94 @@
...
-    num_bytes_evicted = EvictionState_choose_objects_to_evict(
-        eviction_state, plasma_store_info, space_to_free, num_objects_to_evict,
-        objects_to_evict);
+    num_bytes_evicted =
+        choose_objects_to_evict(space_to_free, objects_to_evict);
     LOG_INFO(
         "There is not enough space to create this object, so evicting "
         "%" PRId64 " objects to free up %" PRId64 " bytes.",
-        *num_objects_to_evict, num_bytes_evicted);
+        objects_to_evict.size(), num_bytes_evicted);
```
Comment (robertnishihara) [inline src/plasma/eviction_policy.cc]: This is giving me a compiler warning.

```
/Users/rkn/Workspace/ray/src/plasma/eviction_policy.cc:69:9: warning: format specifies type 'long long' but the argument
      has type 'size_type' (aka 'unsigned long') [-Wformat]
        objects_to_evict.size(), num_bytes_evicted);
~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/rkn/Workspace/ray/src/common/cmake/../common.h:70:67: note: expanded from macro 'LOG_INFO'
  fprintf(stderr, "[INFO] (%s:%d) " M "\n", __FILE__, __LINE__, ##__VA_ARGS__)
                                                                  ^~~~~~~~~~~
```

File: src/plasma/plasma_store.cc (hunk 19 above)
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Hmm.. I just tried out my suggestion and it causes a crash in the plasma tests. It's probably not quite right.
